Libraries/NativeAnimation/RCTNativeAnimatedModule.m

```diff
@@ -162,32 +224,72 @@ - (void)setBridge:(RCTBridge *)bridge
                   eventName:(nonnull NSString *)eventName
                   eventMapping:(NSDictionary<NSString *, id> *)eventMapping)
 {
-  [_operations addObject:^(RCTNativeAnimatedNodesManager *nodesManager) {
+  [self addOperationBlock:^(RCTNativeAnimatedNodesManager *nodesManager) {
     [nodesManager addAnimatedEventToView:viewTag eventName:eventName eventMapping:eventMapping];
   }];
 }
 
 RCT_EXPORT_METHOD(removeAnimatedEventFromView:(nonnull NSNumber *)viewTag
                   eventName:(nonnull NSString *)eventName)
 {
-  [_operations addObject:^(RCTNativeAnimatedNodesManager *nodesManager) {
+  [self addOperationBlock:^(RCTNativeAnimatedNodesManager *nodesManager) {
     [nodesManager removeAnimatedEventFromView:viewTag eventName:eventName];
   }];
 }
 
 #pragma mark -- Batch handling
 
+- (void)addOperationBlock:(AnimatedOperation)operation
+{
+  [_operations addObject:operation];
+}
+
+- (void)addPreOperationBlock:(AnimatedOperation)operation
+{
+  [_preOperations addObject:operation];
+}
+
 - (void)batchDidComplete
 {
-  NSArray *operations = _operations;
+  if (_preOperations.count == 0 && _operations.count == 0) {
+    return;
+  }
+
+  NSArray<AnimatedOperation> *preOperations = _preOperations;
+  NSArray<AnimatedOperation> *operations = _operations;
+  _preOperations = [NSMutableArray new];
   _operations = [NSMutableArray new];
 
-  dispatch_async(dispatch_get_main_queue(), ^{
-    [operations enumerateObjectsUsingBlock:^(AnimatedOperation operation, NSUInteger i, BOOL *stop) {
-      operation(self->_nodesManager);
-    }];
-    [self->_nodesManager updateAnimations];
-  });
+  OperationBatch *batch =
+    [[OperationBatch alloc] initWithNodesManager:_nodesManager
+                                   preOperations:preOperations
+                                      operations:operations];
+
+  [_operationBatchesLock lock];
+  [_operationBatches addObject:batch];
+  [_operationBatchesLock unlock];
+}
+
+- (void)uiManagerWillFlushUIBlocks:(RCTUIManager *)uiManager
+{
+  [_operationBatchesLock lock];
+  [_stagedOperationBatches addObjectsFromArray:_operationBatches];
+  [_operationBatches removeAllObjects];
+  [_operationBatchesLock unlock];
+
+  for (OperationBatch *batch in _stagedOperationBatches) {
```

Comment:
Yea you are right, I checked how UIManager works it it will dispatch to the main queue once per batch and execute only the operations from that batch instead of every it has available. I think what could work well here is only process one batch at a time, this will work because for each completed batch there will be a call to the UIManagerObserver methods.